sway/desktop/transaction.c

```diff
@@ -252,6 +253,413 @@ static void apply_container_state(struct sway_container *container,
+		struct sway_container *child = con->current.children->items[i];
+
+		wlr_scene_node_reparent(child->scene_node, con->content_node);
+
+		disable_container(child);
+	}
+}
+
+static void arrange_container(struct sway_container *con,
+		int width, int height, bool title_bar, int gaps);
+
+static bool arrange_children(enum sway_container_layout layout, list_t *children,
+		struct sway_container *active, struct wlr_scene_node *content,
+		int width, int height, int gaps) {
+
+	int title_bar_height = container_titlebar_height();
+
+	if (layout == L_TABBED) {
+		double w = (double) width / children->length;
+		int title_offset = 0;
+		for (int i = 0; i < children->length; i++) {
+			struct sway_container *child = children->items[i];
+			bool activated = child == active;
+			int next_title_offset = round(w * i + w);
+
+			arrange_title_bar(child, title_offset, -title_bar_height,
+				next_title_offset - title_offset, title_bar_height);
+			wlr_scene_node_set_enabled(child->border.node, activated);
+			wlr_scene_node_set_position(child->scene_node, 0, title_bar_height);
+			wlr_scene_node_reparent(child->scene_node, content);
+
+			if (activated) {
+				arrange_container(child, width, height - title_bar_height,
+					false, 0);
+			} else {
+				disable_container(child);
+			}
+
+			title_offset = next_title_offset;
+		}
+	} else if (layout == L_STACKED) {
+		int title_height = title_bar_height * children->length;
+
+		int y = 0;
+		for (int i = 0; i < children->length; i++) {
+			struct sway_container *child = children->items[i];
+			bool activated = child == active;
+
+			arrange_title_bar(child, 0, y - title_height, width, title_bar_height);
+			wlr_scene_node_set_enabled(child->border.node, activated);
+			wlr_scene_node_set_position(child->scene_node, 0, title_height);
+			wlr_scene_node_reparent(child->scene_node, content);
+
+			if (activated) {
+				arrange_container(child, width, height - title_height,
+					false, 0);
+			} else {
+				disable_container(child);
+			}
+
+			y += title_bar_height;
+		}
+	} else if (layout == L_VERT) {
+		int off = 0;
+		for (int i = 0; i < children->length; i++) {
+			struct sway_container *child = children->items[i];
+			int cheight = child->current.height;
+
+			wlr_scene_node_set_enabled(child->border.node, true);
+			wlr_scene_node_set_position(child->scene_node, 0, off);
+			wlr_scene_node_reparent(child->scene_node, content);
+			arrange_container(child, width, cheight, true, gaps);
+			off += cheight + gaps;
+		}
+	} else if (layout == L_HORIZ) {
+		int off = 0;
+		for (int i = 0; i < children->length; i++) {
+			struct sway_container *child = children->items[i];
+			int cwidth = child->current.width;
+
+			wlr_scene_node_set_enabled(child->border.node, true);
+			wlr_scene_node_set_position(child->scene_node, off, 0);
+			wlr_scene_node_reparent(child->scene_node, content);
+			arrange_container(child, cwidth, height, true, gaps);
+			off += cwidth + gaps;
+		}
+	} else {
+		return false;
```

Comment:
`sway_assert`?

Reply:
The caller is responsible for handling asserts / any additional handling for different layouts.